lib/OpenQA/Worker.pm

```diff
@@ -14,108 +14,336 @@
+    $msg .= "\n - isotovideo version:    " . $self->isotovideo_interface_version;
+    $msg .= "\n - websocket API version: " . WEBSOCKET_API_VERSION;
+    $msg .= "\n - web UI hosts:          " . join(',', @{$settings->webui_hosts});
+    $msg .= "\n - class:                 " . ($settings->global_settings->{WORKER_CLASS} // '?');
+    $msg .= "\n - pool directory:        " . $self->pool_directory;
+    log_info($msg);
+    return $msg;
+}
+
+# determines the worker's capabilities
+sub capabilities {
+    my ($self) = @_;
+
+    if (my $cached_caps = $self->{_caps}) {
+        return $cached_caps;
+    }
+
+    my $global_settings = $self->settings->global_settings;
+    my $query_cmd;
+    my $caps = {
+        host                         => $self->worker_hostname,
+        instance                     => $self->instance_number,
+        websocket_api_version        => WEBSOCKET_API_VERSION,
+        isotovideo_interface_version => $self->isotovideo_interface_version,
+    };
+
+    # determine CPU info
+    if (my $arch = $global_settings->{ARCH}) {
+        $caps->{cpu_arch} = $arch;
+    }
+    else {
+        open(my $LSCPU, "-|", "LC_ALL=C lscpu");
+        for my $line (<$LSCPU>) {
+            chomp $line;
+            if ($line =~ m/Model name:\s+(.+)$/) {
+                $caps->{cpu_modelname} = $1;
+            }
+            if ($line =~ m/Architecture:\s+(.+)$/) {
+                $caps->{cpu_arch} = $1;
+            }
+            if ($line =~ m/CPU op-mode\(s\):\s+(.+)$/) {
+                $caps->{cpu_opmode} = $1;
+            }
+        }
+        close($LSCPU);
+    }
+
+    # determine memory limit
+    open(my $MEMINFO, "<", "/proc/meminfo");
+    for my $line (<$MEMINFO>) {
+        chomp $line;
+        if ($line =~ m/MemTotal:\s+(\d+).+kB/) {
+            my $mem_max = $1 ? $1 : '';
+            $caps->{mem_max} = int($mem_max / 1024) if $mem_max;
+        }
+    }
+    close($MEMINFO);
+
+    # determine worker class ...
+    if (my $worker_class = $global_settings->{WORKER_CLASS}) {
+        # ... from settings
+        $caps->{worker_class} = $worker_class;
+    }
+    else {
+        # ... from CPU architecture
+        my %supported_archs_by_cpu_archs = (
+            i586   => ['i586'],
+            i686   => ['i686', 'i586'],
+            x86_64 => ['x86_64', 'i686', 'i586'],
+
+            ppc     => ['ppc'],
+            ppc64   => ['ppc64le', 'ppc64', 'ppc'],
+            ppc64le => ['ppc64le', 'ppc64', 'ppc'],
+
+            s390  => ['s390'],
+            s390x => ['s390x', 's390'],
+
+            aarch64 => ['aarch64'],
+        );
+        if (my $supported_archs = $supported_archs_by_cpu_archs{$caps->{cpu_arch}}) {
+            $caps->{worker_class} = join(',', map { 'qemu_' . $_ } @$supported_archs);
+            # TODO: check installed qemu and kvm?
+        }
+        else {
+            $caps->{worker_class} = 'qemu_' . $caps->{cpu_arch};
+        }
```

Comment:
how about simply

```perl
$caps->{worker_class} = join(',', map { 'qemu_' . $_ } @$supported_archs_by_cpu_archs{$caps->{cpu_arch}} // $caps->{cpu_arch};
```

This IMHO makes the intention clearer: You want to have a valid worker class and get it from the hash with a default fallback. This is what `//` is used for. As you check if the temporary assigned variable is defined anyway that con be combined. This is only a little bit long because of the long name of the outer hash ;) What it does can be read as:

```perl
$worker_class = join(',', map { 'qemu_' . $_ } @$s{$arch} // $arch;
```